### PR TITLE
Add detailed diagnostics for E0383.

### PR DESCRIPTION
**Pull request by nathankleyn**
This adds detailed diagnostics for `E0383`, 'partial reinitialization of
uninitialized structure'.

This is part of rust-lang/rust#24407.

r? @Manishearth

### CODE REVIEW

File: src/librustc_borrowck/diagnostics.rs [hunk 1]
```diff
@@ -138,6 +138,30 @@ Book:
...
 
+E0383: r##"
+This error occurs when an attempt is made to partially reinitialize a
+structure that is currently uninitialized.
+
+For example, this can happen when a transfer of ownership has taken place:
+
+```
+let mut t = Test { a: 1, b: None};
+let mut u = Test { a: 2, b: Some(Box::new(t))}; // `t` is now uninitialized
```
Comment (Manishearth) [inline src/librustc_borrowck/diagnostics.rs]: Can we just `drop(t)` or move `t` elsewhere here (e.g. `let u = Box::new(t)`)? The recursive type adds to the confusion.

**Comment by Manishearth** [comment]
Thanks for the PR! This looks good, just some minor changes to the code example would be needed.

**Comment by nathankleyn** [comment]
@Manishearth Done and ready for re-review!

r? @Manishearth

**Comment by Manishearth** [comment]
Thanks!

**Pull request merged**